File: schemas/video_script.py
"""Pipeline contract (spec §6). Source of truth.

Mirror: remotion/src/types/video-script.ts (zod). Shared fixture:
schemas/fixtures/sample_video_script.json is validated by both sides.
"""

from typing import Literal

from pydantic import BaseModel, Field, model_validator

VisualType = Literal["ai_broll", "ai_image", "screen_recording", "text_card"]
TemplateName = Literal["explainer", "tutorial", "listicle", "comparison"]

AI_VISUAL_TYPES: frozenset[str] = frozenset({"ai_broll", "ai_image"})


class Segment(BaseModel):
    """One narrated beat of the video with its visual."""

    id: str = Field(min_length=1)
    text: str = Field(min_length=1)
    visual_type: VisualType
    visual_prompt: str | None = None
    duration_estimate_s: float = Field(gt=0)
    caption_emphasis: list[str] = Field(default_factory=list)

    @model_validator(mode="after")
    def _ai_visuals_need_prompt(self) -> "Segment":
        if self.visual_type in AI_VISUAL_TYPES and not self.visual_prompt:
            raise ValueError(
                f"visual_prompt is required when visual_type={self.visual_type!r}"
            )
        return self


class VideoScript(BaseModel):
    """Everything downstream stages need to produce one video."""

    topic: str = Field(min_length=1)
    template: TemplateName
    hook: Segment
    segments: list[Segment] = Field(min_length=1)
    cta: Segment
    target_duration_s: int = Field(ge=30, le=60)
    platform_captions: dict[str, str]
    hashtags: dict[str, list[str]]

    @model_validator(mode="after")
    def _segment_ids_unique(self) -> "VideoScript":
        ids = [s.id for s in self.all_segments()]
        if len(ids) != len(set(ids)):
            raise ValueError("segment ids must be unique across hook, segments, and cta")
        return self

    def all_segments(self) -> list[Segment]:
        """Hook, body segments, CTA — in playback order."""
        return [self.hook, *self.segments, self.cta]


class WordTiming(BaseModel):
    """One word from ElevenLabs word-level timestamps (caption sync)."""

    word: str
    start_s: float = Field(ge=0)
    end_s: float

    @model_validator(mode="after")
    def _end_after_start(self) -> "WordTiming":
        if self.end_s < self.start_s:
            raise ValueError("end_s must be >= start_s")
        return self


class VideoAssets(BaseModel):
    """Shape of videos.asset_urls jsonb. Keys are segment ids."""

    voiceover: dict[str, str] = Field(default_factory=dict)
    visuals: dict[str, str] = Field(default_factory=dict)
    timings: dict[str, list[WordTiming]] = Field(default_factory=dict)
    thumbnail_base: str | None = None

File: remotion/src/types/video-script.ts
/**
 * Mirror of schemas/video_script.py (the source of truth — spec §6).
 * Change Python first, then this file, then the shared fixtures.
 */
import {z} from 'zod';

export const AI_VISUAL_TYPES = ['ai_broll', 'ai_image'] as const;

export const segmentSchema = z
  .object({
    id: z.string().min(1),
    text: z.string().min(1),
    visual_type: z.enum(['ai_broll', 'ai_image', 'screen_recording', 'text_card']),
    // null means "no prompt needed" (text_card / screen_recording); the key may
    // also be absent entirely (Python defaults it to None, and exclude_none dumps drop it).
    visual_prompt: z.string().nullable().optional(),
    duration_estimate_s: z.number().positive(),
    caption_emphasis: z.array(z.string()),
  })
  .refine(
    (s) =>
      !(AI_VISUAL_TYPES as readonly string[]).includes(s.visual_type) ||
      (s.visual_prompt ?? '').trim().length > 0,
    {message: 'visual_prompt is required for ai_broll/ai_image segments', path: ['visual_prompt']},
  );

export const videoScriptSchema = z
  .object({
    topic: z.string().min(1),
    template: z.enum(['explainer', 'tutorial', 'listicle', 'comparison']),
    hook: segmentSchema,
    segments: z.array(segmentSchema).min(1).max(5),
    cta: segmentSchema,
    target_duration_s: z.number().int().min(30).max(60),
    platform_captions: z.record(z.string()),
    hashtags: z.record(z.array(z.string())),
  })
  .refine(
    (s) => {
      const ids = [s.hook.id, ...s.segments.map((x) => x.id), s.cta.id];
      return new Set(ids).size === ids.length;
    },
    {message: 'segment ids must be unique across hook, segments, and cta'},
  );

export const wordTimingSchema = z
  .object({
    word: z.string().min(1),
    start_s: z.number().min(0),
    end_s: z.number(),
  })
  .refine((w) => w.end_s >= w.start_s, {message: 'end_s must be >= start_s'});

// Defaults to {} so partial asset objects (e.g. voiceover-only) parse successfully.
// Render compositions must guard lookups: const words = assets.timings[segId] ?? [];
export const videoAssetsSchema = z.object({
  voiceover: z.record(z.string()).default({}),
  visuals: z.record(z.string()).default({}),
  timings: z.record(z.array(wordTimingSchema)).default({}),
  thumbnail_base: z.string().nullable().default(null),
});

export type Segment = z.infer<typeof segmentSchema>;
export type VideoScript = z.infer<typeof videoScriptSchema>;
export type WordTiming = z.infer<typeof wordTimingSchema>;
export type VideoAssets = z.infer<typeof videoAssetsSchema>;

/** Hook, body segments, CTA — in playback order. */
export const allSegments = (script: VideoScript): Segment[] => [
  script.hook,
  ...script.segments,
  script.cta,
];
